python/perf-kernels/rmsnorm.py

```diff
@@ -37,100 +46,96 @@ def get_autotune_config():
         return get_hip_autotune_config()
 
 
-# accumulate sum of squares for a row in a blocked manner
-@triton.jit
-def accumulate_sum_squares(input_ptr, input_row_stride, n_cols, BLOCK_SIZE, row_idx):
-    col_offsets = tl.arange(0, BLOCK_SIZE)
-    sum_squares = tl.zeros([1], dtype=tl.float32)
-    row_input_ptr = input_ptr + row_idx * input_row_stride
-
-    n_cols_blks = tl.cdiv(n_cols, BLOCK_SIZE) - 1
-    for start in range(0, n_cols_blks * BLOCK_SIZE, BLOCK_SIZE):
-        cols = start + col_offsets
-        input_ptrs = row_input_ptr + cols
-        input_ptrs = tl.multiple_of(input_ptrs, (16, ))
-        x = tl.load(input_ptrs)
-        sum_squares += tl.sum(x * x, axis=0)
-
-    # loop peeling for mask
-    cols = n_cols_blks * BLOCK_SIZE + col_offsets
-    mask = cols < n_cols
-    input_ptrs = row_input_ptr + cols
-    input_ptrs = tl.multiple_of(input_ptrs, (16, ))
-    x = tl.load(input_ptrs, mask=mask, other=0.0, cache_modifier=".cg")
-    sum_squares += tl.sum(x * x, axis=0)
-
-    return sum_squares
-
-
-# apply normalization to each block of the row
-@triton.jit
-def apply_normalization(input_ptr, output_ptr, g_ptr, input_row_stride, output_row_stride, n_cols, norm_factor,
-                        BLOCK_SIZE, row_idx):
-    col_offsets = tl.arange(0, BLOCK_SIZE)
-    row_input_ptr = input_ptr + row_idx * input_row_stride
-    row_output_ptr = output_ptr + row_idx * output_row_stride
-
-    for start in range(0, n_cols, BLOCK_SIZE):
-        cols = start + col_offsets
-        mask = cols < n_cols
-        input_ptrs = row_input_ptr + cols
-        input_ptrs = tl.multiple_of(input_ptrs, (16, ))
-        g_ptrs = g_ptr + cols
-        output_ptrs = row_output_ptr + cols
-        x = tl.load(input_ptrs, mask=mask, other=0.0, cache_modifier=".cg")
-        g = tl.load(g_ptrs, mask=mask, other=0.0)
-        rms_norm = x * norm_factor * g
-        tl.store(output_ptrs, rms_norm, mask=mask)
-
-
-# Main kernel with both blocked and non-blocked versions based on BLOCK_SIZE
 @triton.autotune(configs=get_autotune_config(), key=['n_rows', 'n_cols'], use_cuda_graph=True)
 @triton.jit
 def rms_kernel(output_ptr, input_ptr, g_ptr, input_row_stride, output_row_stride, n_rows, n_cols, epsilon,
                BLOCK_SIZE: tl.constexpr, USE_BLOCKED: tl.constexpr, NUM_PRGMS: tl.constexpr):
-    row_idx = tl.program_id(0)  # Each program instance handles one row
+    row_start = tl.program_id(0)  # Each program instance handles one row
     col_offsets = tl.arange(0, BLOCK_SIZE)
+    tl.assume(input_row_stride >= 0)
+    tl.assume(output_row_stride >= 0)
 
     if USE_BLOCKED:
```

Comment:
There are no common bits to the two implementations anymore. Does it make more sense to call different implementations each with its own kernel impl? Or have a single large kernel like now?

Reply:
With single kernel implementation, using 
```
BLOCK_SIZE = blk_size
USE_BLOCKED = n_cols > BLOCK_SIZE
```

and 

```
MAX_FUSED_SIZE = 65536 // x.element_size()
blk_size = min(MAX_FUSED_SIZE, triton.next_power_of_2(n_cols))
```

we can automatically switch between blocked and non blocked